### PR TITLE
100 integration testing with testcontainers

### PR DESCRIPTION
**Pull request by andregasser**
The changes in this PR provide abasis for running integration tests against different versions of Mastodon. I have provided setups for Mastodon 3.0.0, 4.0.2 and 4.1.0. More could be added if needed. Setups are based on Docker Compose. For more information on the setup or on how to run these tests, please check out https://github.com/andregasser/bigbone/wiki/Testing

This PR closes https://github.com/andregasser/bigbone/issues/100

### CODE REVIEW

**Comment by andregasser** [comment]
@bocops Would be great if you could have a look at these changes. More details about how to test our code can be found here: https://github.com/andregasser/bigbone/wiki/Testing.

File: bigbone/build.gradle [hunk 1]
```diff
@@ -2,13 +2,48 @@ plugins {
...
 
-    testImplementation libs.junit.jupiter
+    testImplementation libs.junit.jupiter.api
     testImplementation libs.kluent
     testImplementation libs.mockk
     testImplementation libs.mockk.dsl
+    integrationTestImplementation libs.junit.jupiter.api
+    integrationTestImplementation libs.junit.jupiter.engine
+    integrationTestImplementation libs.junit.platform.suite.engine
+    integrationTestImplementation libs.testcontainers
```
Comment (andregasser) [inline bigbone/build.gradle:29]: testcontainers dependencies are not needed atm, as we just rely on pure docker compose for the moment.... Will clean this up.

File: bigbone-rx/src/main/kotlin/social/bigbone/rx/RxOAuthMethods.kt [hunk 2]
```diff
@@ -15,32 +15,49 @@ import social.bigbone.rx.extensions.onErrorIfNotDisposed
 class RxOAuthMethods(client: MastodonClient) {
     private val oauth = OAuthMethods(client)
 
-    fun getAccessTokenWithAuthorizationCodeGrant(
+    fun getClientAccessToken(
```
Comment (andregasser) [inline bigbone-rx/src/main/kotlin/social/bigbone/rx/RxOAuthMethods.kt:18]: You have implemented the same method in one of your other pull requests. I'll have a look and see which and how we merge this.

File: bigbone/src/integrationTest/resources/logback.xml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<configuration debug="true">
```
Comment (andregasser) [inline bigbone/src/integrationTest/resources/logback.xml:1]: Will remove this file from the PR as this was only needed for testcontainers.

File: docker/v402/.env [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+SECRET_KEY_BASE=gfhgfhgfhhtrhtr
+OTP_SECRET=hthtrzhtrh
+
+VAPID_PRIVATE_KEY=OdpiFjiFZmmZZvC5-GBWWrUAiTGHQ4GvHT1vxIz5vww=
+VAPID_PUBLIC_KEY=BGw1VsPbWjP6umtRSmdgoz881u-eB8MIw8tPCxJREC6yhwWsqpcWziPXoIVussPj4gcvSyQNFv8d_TEBBdGcRUQ=
+
+S3_ENABLED=false
+
+
+#
```
Comment (andregasser) [inline docker/v402/.env:27]: Will remove these comments.

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -5,12 +5,15 @@ bigbone = "2.0.0-SNAPSHOT"
...
 gson = "2.10.1"
 junit-jupiter = "5.9.2"
+junit-platform-suite-engine = "1.9.2"
 kluent = "1.72"
 kotlin = "1.7.21"
 kotlin-coroutines = "1.6.4"
 mockk = "1.13.4"
 okhttp = "4.10.0"
 rxjava = "3.1.6"
+testcontainers = "1.17.6"
```
Comment (andregasser) [inline gradle/libs.versions.toml:15]: Will remove these testcontainers dependencies as well.

File: bigbone/build.gradle (hunk 1 above)
Comment (andregasser) [reply bigbone/build.gradle]: Fixed.

File: bigbone/src/integrationTest/resources/logback.xml (hunk 3 above)
Comment (andregasser) [reply bigbone/src/integrationTest/resources/logback.xml]: Fixed.

File: gradle/libs.versions.toml (hunk 5 above)
Comment (andregasser) [reply gradle/libs.versions.toml]: Fixed.

File: bigbone/src/main/kotlin/social/bigbone/api/method/OAuthMethods.kt [hunk 6]
```diff
@@ -24,18 +24,49 @@ class OAuthMethods(private val client: MastodonClient) {
...
+     * @param clientId The client ID, obtained during app registration.
+     * @param clientSecret The client secret, obtained during app registration.
+     * @param scope List of requested OAuth scopes, separated by spaces. Must be a subset of scopes declared during app registration.
+     * @param redirectUri Set a URI to redirect the user to. Defaults to "urn:ietf:wg:oauth:2.0:oob",
+     *  which will display the authorization code to the user instead of redirecting to a web page.
+     *  Must match one of the redirect_uris declared during app registration.
+     *  @see <a href="https://docs.joinmastodon.org/methods/oauth/#token">Mastodon oauth API methods #token</a>
+     */
+    @JvmOverloads
+    fun getClientAccessToken(
```
Comment (andregasser) [inline bigbone/src/main/kotlin/social/bigbone/api/method/OAuthMethods.kt:49]: I will undo this implementation of OAuth2 password grant flow, as you have already implemented it in https://github.com/andregasser/bigbone/pull/147.

File: bigbone-rx/src/main/kotlin/social/bigbone/rx/RxOAuthMethods.kt (hunk 2 above)
Comment (andregasser) [reply bigbone-rx/src/main/kotlin/social/bigbone/rx/RxOAuthMethods.kt]: Done.

File: bigbone/src/main/kotlin/social/bigbone/api/method/OAuthMethods.kt (hunk 6 above)
Comment (andregasser) [reply bigbone/src/main/kotlin/social/bigbone/api/method/OAuthMethods.kt]: Done.

File: docker/v402/.env (hunk 4 above)
Comment (andregasser) [reply docker/v402/.env]: Done.

File: bigbone/src/integrationTest/kotlin/social/bigbone/TestConstants.kt [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package social.bigbone
+
+class TestConstants {
```
Comment (bocops) [inline bigbone/src/integrationTest/kotlin/social/bigbone/TestConstants.kt:3]: Some of these constants seem unused. Especially, port definitions (443 for both REST and streaming APIs) are not used and different from the ports mentioned on the Testing wiki page (3000/4000). Is this correct?

File: bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt [hunk 8]
```diff
@@ -417,11 +432,32 @@ private constructor(
...
         private var debug = false
+        private var scheme = "https"
+        private var port = 443
+        private var trustAllCerts = false
 
         fun accessToken(accessToken: String) = apply {
             this.accessToken = accessToken
         }
 
+        fun withHttpsDisabled() = apply {
```
Comment (bocops) [inline bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt:443]: Can public Mastodon servers run on an HTTP scheme? If not, we should at least advise against using this in production, similar to the documentation of `withTrustAllCerts()`.

File: bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt [hunk 9]
```diff
@@ -417,11 +432,32 @@ private constructor(
...
+         * Disables certificate validation and hostname verification. Please do not use this
+         * on production environments, as it is considered bad practice. Use it solely for
+         * testing purposes.
+         */
+        fun withTrustAllCerts() = apply {
+            trustAllCerts = true
+            configureForTrustAll(okHttpClientBuilder)
+        }
+
+        fun withPort(port: Int) = apply {
```
Comment (bocops) [inline bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt:457]: Can public Mastodon servers run on ports other than 443? If not, we should at least advise against using this in production, similar to the documentation of `withTrustAllCerts()`.

**Review by bocops — COMMENTED** [review]
I couldn't immediately get Docker and the integration tests to run, but that might be a problem on my end. I got a "FATAL: database 'mastodon' does not exist" and then an exit with code 1 " container for service 'web' is unhealthy shortly thereafter. Looking into it.

Other than that, most of this looks good, I only have some small questions about the test constants and the scheme/port configuration changes to `MastodonClient.Builder`.

File: bigbone/src/integrationTest/kotlin/social/bigbone/TestConstants.kt (hunk 7 above)
Comment (andregasser) [reply bigbone/src/integrationTest/kotlin/social/bigbone/TestConstants.kt:3]: Ye, there are some unused constants. I'll double check. The way the Docker Compose setup works is basically this:
- Mastodon REST instance on port 3000 is started
- Mastodon Streaming instance on port 4000 is started
- Nginx is launched which accepts traffic on port 443 and forwards it to one of them above.

I've decided to go this way, as the recommendation for running Ruby locally is using SSL, as some stuff might not work properly if served via HTTP only. That's why Nginx is used as reverse proxy in front of them.

If a client requests something like /api/v2/streaming/... it goes to the streaming instance, anything else goes to the regular instance.

So for the integration tests, we should connect MastodonClient to port 443 for the moment. We can change this of course.

**Comment by andregasser** [comment]
> I couldn't immediately get Docker and the integration tests to run, but that might be a problem on my end. I got a "FATAL: database 'mastodon' does not exist" and then an exit with code 1 " container for service 'web' is unhealthy shortly thereafter. Looking into it.
> 
> Other than that, most of this looks good, I only have some small questions about the test constants and the scheme/port configuration changes to `MastodonClient.Builder`.

Sounds a bit like the 'db' container is not running correctly. When 'db' container is created, a user with DB-creation rights is created, but the database itself is created in the 'web' container using `db:setup` call.

you can send me your docker compose logs if you want then I'll have a look too.

File: bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt (hunk 8 above)
Comment (andregasser) [reply bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt:443]: For production setups this should clearly not be used. Back then, I have added it for testing-purposes only. My plan was to run local Mastodon instances on HTTP only (for simplicity reasons). But I've soon realized, that Ruby on Rails together with Mastodon code works best using HTTPS (saw a comment somewhere, people claiming that not everything might work as expected, if stuff is running on HTTP). So I've built the test setup based on HTTPS with certificates. 

Now, as the HTTPS setup is working, we could simply kill this method.

**Comment by andregasser** [comment]
Regarding the following methods:
- `withTrustAllCerts()`
- `withPort()`
- `withHttpsDisabled()`

Clearly, `withTrustAllCerts()` and `withHttpsDisabled` should never be used in production setups (because they have a security impact). Ideally, a Mastodon instance always runs on the default SSL port 443, but I honestly have no clue whether this is always the case (certainly it would be good practice). 

Maybe these methods still could be handy / useful for people who want to test out stuff. 

So would it be ok for you, if we add a clear KDoc reminder to not use these methods in production setups?
Also, we could add a log.WARN in the code to advice against using this on prod.

Additional note for 'withPort()`: At the moment, the local Mastodon instance running on Docker Compose is reachable via port 443, so in your browser you can just enter `https://localhost` and you end up on the login screen of the Mastodon instance. While this is comfortable, I am not sure, if it is clever to occupy the 443 port locally, There are probably quite some devs that use this port to run some other web based services locally, That would cause port conflicts. So, if we switch to another port (e.g. 5000, we required this method). 

Do you have any preference on how to proceed?

**Comment by bocops** [comment]
Sounds good - clear KDoc warnings should be good enough. :)

**Pull request merged**